File: lab01/builder/src/main/java/com/example/builder/KnightBuilder.java
package com.example.builder;

import com.example.builder.enums.CharacterType;
import com.example.builder.enums.WeaponType;

public class KnightBuilder implements CharacterBuilder {

    private String name;
    private Integer health;
    private Integer mana;
    private WeaponType weapon;

    @Override
    public CharacterBuilder reset() {
        name = null;
        health = null;
        mana = null;
        weapon = null;
        return this;
    }

    @Override
    public CharacterBuilder setName(String name) {
        this.name = name;
        return this;
    }

    @Override
    public CharacterBuilder setHealth(Integer health) {
        this.health = health;
        return this;
    }

    @Override
    public CharacterBuilder setMana(Integer mana) {
        this.mana = mana;
        return this;
    }

    @Override
    public CharacterBuilder setWeapon(WeaponType weapon) {
        this.weapon = weapon;
        return this;
    }

    @Override
    public Character build() {
        return new Character(CharacterType.KNIGHT, name, health, mana, weapon);
    }
    
}

File: lab01/builder1/Builders/KnightBuilder.cs
using Builder1.Enums;
using Builder1.Models;

namespace Builder1.Builders;

public class KnightBuilder : ICharacterBuilder
{
    private string? _name;
    private int? _health;
    private int? _mana;
    private WeaponType? _weapon;

    public ICharacterBuilder Reset()
    {
        _name = null;
        _health = null;
        _mana = null;
        _weapon = null;
        return this;
    }

    public ICharacterBuilder SetName(string name) { _name = name; return this; }
    public ICharacterBuilder SetHealth(int? health) { _health = health; return this; }
    public ICharacterBuilder SetMana(int? mana) { _mana = mana; return this; }
    public ICharacterBuilder SetWeapon(WeaponType? weapon) { _weapon = weapon; return this; }

    public Character Build() => new(CharacterType.Knight, _name!, _health, _mana, _weapon);
}
